In Class April 9th

In [9]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [10]:
# Importing Libraries
import pandas as pd
import numpy as np
import sweetviz as sv
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, accuracy_score

In [11]:
# Importing the dataset
default = pd.read_csv('creditcard.csv')
default.head()

,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6,BILL_AMT1,BILL_AMT2,BILL_AMT3,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default payment next month
0,400000,1,1,2,32,0,0,0,0,0,0,55773,55917,51389,48272,49478,51242,3028,3023,3000,3000,3000,38662,0
1,120000,2,2,2,30,-1,-1,-1,-1,-1,-1,140,3230,3011,1964,1883,1538,3230,3011,1964,1883,1538,1911,0
2,270000,2,2,2,32,0,0,0,0,0,0,59710,49986,104390,94856,86461,83650,1808,69563,2891,2689,3012,2771,0
3,280000,2,2,1,27,0,0,0,0,0,0,280913,283222,273160,257689,193231,191143,11052,9563,15017,5374,5420,6021,0
4,30000,2,1,2,27,0,0,-1,0,0,-2,1512,2458,664,1814,0,0,1000,664,1500,0,0,0,0


In [12]:
# Initial EDA- this is not a balanced dataset
report = sv.analyze(default)
report.show_html('sweet_report.html')

                                             |          | [  0%]   00:00 -> (? left)

Report sweet_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


Highlights from EDA: 
There is no missing values in the dataset. There are also a mix of features (categorical and numerical). I noticed that there are multiple financial variables, such as BILL_AMTs that are skewed right. 


In [13]:
# baseline rd and xgboost models
# Prepare the data for modeling
X = default.drop('default payment next month', axis=1)
y = default['default payment next month']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, shuffle=True, random_state=42)
# Train the random forest model
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)
# Evaluate the random forest model
y_pred_rf = rf.predict(X_test)
print("Random Forest Classification Report:")
print(classification_report(y_test, y_pred_rf))
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
# Train the XGBoost model
xgb = XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')
xgb.fit(X_train, y_train)
# Evaluate the XGBoost model
y_pred_xgb = xgb.predict(X_test)
print("XGBoost Classification Report:")
print(classification_report(y_test, y_pred_xgb))
print("XGBoost Accuracy:", accuracy_score(y_test, y_pred_xgb))



Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.93      0.89      3738
           1       0.62      0.38      0.47      1062

    accuracy                           0.81      4800
   macro avg       0.73      0.66      0.68      4800
weighted avg       0.79      0.81      0.79      4800

Random Forest Accuracy: 0.81125
XGBoost Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.93      0.88      3738
           1       0.61      0.38      0.47      1062

    accuracy                           0.81      4800
   macro avg       0.73      0.66      0.68      4800
weighted avg       0.79      0.81      0.79      4800

XGBoost Accuracy: 0.8102083333333333


/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:09:48] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


In [14]:
# building baseline models with cross validation
rf_cv_scores = cross_val_score(rf, X, y, cv=5)
xgb_cv_scores = cross_val_score(xgb, X, y, cv=5)
print("Random Forest CV Scores:", rf_cv_scores)
print("Random Forest CV Mean Score:", np.mean(rf_cv_scores))
print("XGBoost CV Scores:", xgb_cv_scores)
print("XGBoost CV Mean Score:", np.mean(xgb_cv_scores))

#printing the standard deviations
print("Random Forest CV Mean:", np.mean(rf_cv_scores))
print("Random Forest CV Std:", np.std(rf_cv_scores))

print("XGBoost CV Mean:", np.mean(xgb_cv_scores))
print("XGBoost CV Std:", np.std(xgb_cv_scores))

/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:06] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:06] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:07] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:07] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i

Random Forest CV Scores: [0.81125    0.81625    0.814375   0.80958333 0.81875   ]
Random Forest CV Mean Score: 0.8140416666666667
XGBoost CV Scores: [0.80979167 0.81083333 0.814375   0.81020833 0.81895833]
XGBoost CV Mean Score: 0.8128333333333334
Random Forest CV Mean: 0.8140416666666667
Random Forest CV Std: 0.0033113860810509168
XGBoost CV Mean: 0.8128333333333334
XGBoost CV Std: 0.00346510381438197


/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:07] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Both RF and XGBoost show very similar performance so far, but RF slightly outperforms XGBoost in terms of accuracy. The standard deviations are also very low (about 0.003) for both models. The different between the two models' performances are minimal so far. The mean score represents the average performance across the five folds, and the standard deviation indicates how consistent the model is across different subsets of data. 

In [15]:
# repeat baseline model training but use recall as teh evaluation metrics
rf_cv_recall_scores = cross_val_score(rf, X, y, cv=5, scoring='recall')
xgb_cv_recall_scores = cross_val_score(xgb, X, y, cv=5, scoring='recall')
print("Random Forest CV Recall Scores:", rf_cv_recall_scores)
print("Random Forest CV Mean Recall Score:", np.mean(rf_cv_recall_scores))
print("XGBoost CV Recall Scores:", xgb_cv_recall_scores)
print("XGBoost CV Mean Recall Score:", np.mean(xgb_cv_recall_scores))   
print(classification_report(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_xgb))


/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:25] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:25] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:26] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:26] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i

Random Forest CV Recall Scores: [0.35344015 0.36629002 0.36440678 0.36252354 0.37099812]
Random Forest CV Mean Recall Score: 0.3635317213090021
XGBoost CV Recall Scores: [0.35626767 0.37288136 0.37193974 0.36817326 0.37382298]
XGBoost CV Mean Recall Score: 0.36861699956158334
              precision    recall  f1-score   support

           0       0.84      0.93      0.89      3738
           1       0.62      0.38      0.47      1062

    accuracy                           0.81      4800
   macro avg       0.73      0.66      0.68      4800
weighted avg       0.79      0.81      0.79      4800

              precision    recall  f1-score   support

           0       0.84      0.93      0.88      3738
           1       0.61      0.38      0.47      1062

    accuracy                           0.81      4800
   macro avg       0.73      0.66      0.68      4800
weighted avg       0.79      0.81      0.79      4800



/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:26] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Although both models seem to perform pretty well with accuracy, I see that recall is only about 38%. Looking at accuracy alone would be misleading when evaluating the model performance. This may be due to the class imbalance that we saw from the initial EDA report. 

In [16]:
# rebuild rf and xgboost models with adjustment to class weights to address class imbalance
rf_weighted = RandomForestClassifier(random_state=42, class_weight='balanced')
rf_weighted.fit(X_train, y_train)
y_pred_rf_weighted = rf_weighted.predict(X_test)
print("Random Forest with Class Weights Classification Report:")
print(classification_report(y_test, y_pred_rf_weighted))
xgb_weighted = XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss', scale_pos_weight= (y_train == 0).sum() / (y_train == 1).sum())

xgb_weighted.fit(X_train, y_train)
y_pred_xgb_weighted = xgb_weighted.predict(X_test)
print("XGBoost with Class Weights Classification Report:")
print(classification_report(y_test, y_pred_xgb_weighted))


Random Forest with Class Weights Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.94      0.89      3738
           1       0.64      0.36      0.46      1062

    accuracy                           0.81      4800
   macro avg       0.74      0.65      0.67      4800
weighted avg       0.79      0.81      0.79      4800



/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:10:29] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost with Class Weights Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.82      0.84      3738
           1       0.47      0.58      0.52      1062

    accuracy                           0.76      4800
   macro avg       0.67      0.70      0.68      4800
weighted avg       0.79      0.76      0.77      4800



RF model performance got slightly worse when it comes to recall (0.38 --> 0.36), but saw no major difference. XGBoost saw a big improvement (0.38 --> 0.58) in recall. However, accuracy and precision did go down. This shows that there are tradeoffs. 

In [17]:
# hyperparameter tuning with grid search for random forest with 2 hyperparameters and 2 values for each
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10]
}
grid_search_rf = GridSearchCV(estimator=rf_weighted, param_grid=param_grid_rf, cv=5, scoring='recall')
grid_search_rf.fit(X_train, y_train)
print("Best Hyperparameters for Random Forest:", grid_search_rf.best_params_)
best_rf = grid_search_rf.best_estimator_
y_pred_best_rf = best_rf.predict(X_test)
print("Best Random Forest Classification Report:")
print(classification_report(y_test, y_pred_best_rf))    

Best Hyperparameters for Random Forest: {'max_depth': 10, 'n_estimators': 200}
Best Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.85      0.86      3738
           1       0.51      0.56      0.53      1062

    accuracy                           0.78      4800
   macro avg       0.69      0.70      0.70      4800
weighted avg       0.79      0.78      0.79      4800



In [18]:
# hyperparameter tuning with grid search for xgboost with 2 hyperparameters and 2 values for each
from pytest import param


param_grid_xgb = {
    'n_estimators': [100, 200],
    'max_depth': [3, 6]
}
grid_search_xgb = GridSearchCV(estimator=xgb_weighted, param_grid=param_grid_xgb, cv=5, scoring='recall')
grid_search_xgb.fit(X_train, y_train)
print("Best Hyperparameters for XGBoost:", grid_search_xgb.best_params_)
best_xgb = grid_search_xgb.best_estimator_
y_pred_best_xgb = best_xgb.predict(X_test)
print("Best XGBoost Classification Report:") 
print(classification_report(y_test, y_pred_best_xgb))     

/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:11:43] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:11:43] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:11:43] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:11:43] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i

Best Hyperparameters for XGBoost: {'max_depth': 3, 'n_estimators': 100}
Best XGBoost Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.79      0.83      3738
           1       0.46      0.64      0.54      1062

    accuracy                           0.76      4800
   macro avg       0.67      0.72      0.69      4800
weighted avg       0.79      0.76      0.77      4800



/opt/anaconda3/lib/python3.13/site-packages/xgboost/training.py:200: UserWarning: [17:11:46] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


After tuning, the models seemed to both improve. For RF, recall improved from 0.36 to 0.56. F1 improved from 0.46 to 0.53. For XG, recall improved from 0.58 to 0.64 and F1 improved from 0.52 to 0.54. Tuning the XG boost did make it improve, but not as much of a difference as seen in the RF model. I think there are so many different parameters to consider and different values to use to tune in order to find the best model. 

I am interpreting "out-of-the-box" to mean one of the very first original models before tuning and using weights. RF had a slightly higher accuracy score (RF:0.814 and XG:0.813), and the classification report showed that the recalls were also virtually the same (RF:0.38 and XG:0.38). Performance is very similar, but random forest performed slightly better. 